### PR TITLE
Set role

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 1]
```diff
@@ -753,6 +754,21 @@ public Node visitRevokeRoles(SqlBaseParser.RevokeRolesContext context)
...
 
     @Override
+    public Node visitSetRole(SqlBaseParser.SetRoleContext context)
+    {
+        Optional<String> catalog = getTextIfPresent(context.catalog);
+        Optional<String> role = getTextIfPresent(context.role);
+        SetRole.Type type = SetRole.Type.ROLE;
+        if (context.ALL() != null) {
+            type = SetRole.Type.ALL;
+        }
```
Comment (cawallin) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:764]: nit: the code style dictates no newline here before the else if

File: presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java [hunk 2]
```diff
@@ -554,6 +554,18 @@ public void checkCanRevokeRoles(TransactionId transactionId, Identity identity,
         }
     }
 
+    @Override
+    public void checkCanSetRole(TransactionId transactionId, Identity identity, String role, String catalogName)
+    {
+        requireNonNull(identity, "identity is null");
+        requireNonNull(role, "role is null");
+        requireNonNull(catalogName, "catalog is null");
+        CatalogAccessControlEntry entry = getConnectorAccessControl(transactionId, catalogName);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java:563]: Add this for SystemAccessControl too --  I could see a case where you have an admin role in multiple catalogs and want to be able to only set that role with a given Presto user/list of users.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java [hunk 3]
```diff
@@ -124,17 +129,19 @@ private InternalTable getInternalTable(ConnectorTransactionHandle transactionHan
         InformationSchemaTableHandle handle = split.getTableHandle();
         Map<String, NullableValue> filters = split.getFilters();
 
+        Identity identity = connectorSession.getIdentity();
         Session session = Session.builder(metadata.getSessionPropertyManager())
-                .setTransactionId(transaction.getTransactionId())
                 .setQueryId(new QueryId(connectorSession.getQueryId()))
-                .setIdentity(connectorSession.getIdentity())
+                .setIdentity(new Identity(identity.getUser(), identity.getPrincipal()))
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:135]: Why is this necessary?

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java [hunk 4]
```diff
@@ -124,17 +129,19 @@ private InternalTable getInternalTable(ConnectorTransactionHandle transactionHan
...
                 .setSource("information_schema")
                 .setCatalog("") // default catalog is not be used
                 .setSchema("") // default schema is not be used
                 .setTimeZoneKey(connectorSession.getTimeZoneKey())
                 .setLocale(connectorSession.getLocale())
                 .setStartTime(connectorSession.getStartTime())
-                .build();
+                .setRole(handle.getCatalogName(), identity.getRole())
+                .build()
+                .beginTransactionId(transaction.getTransactionId(), transactionManager, accessControl);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:144]: Please mention in the commit message why this is needed and/or add a separate commit for it -- is this to check setRole access control permissions? What is the UX if it fails here?

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 5]
```diff
@@ -396,6 +399,40 @@ public static RoleGrant fromRolePrincipalGrant(RolePrincipalGrant grant)
...
+    {
+        Optional<SelectedRole> role = identity.getRole();
+
+        if (role.isPresent() && role.get().getType() == SelectedRole.Type.NONE) {
+            return ImmutableSet.of("public");
+        }
+
+        PrestoPrincipal principal;
+        if (!role.isPresent() || role.get().getType() == SelectedRole.Type.ALL) {
+            principal = new PrestoPrincipal(USER, identity.getUser());
```
Comment (cawallin) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:412]: So basically, ALL for the logged in user is the default, just like in Hive -- which doesn't match the standard but matches the Hive behavior, which is more important.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -2054,6 +2055,112 @@ public void testGrantRevokeNonExistingRoles()
...
+                .setRole("hive", new SelectedRole(SelectedRole.Type.ROLE, Optional.of("set_role_3")))
+                .build();
+        Session setRole4 = Session.builder(queryRunner.getDefaultSession())
+                .setIdentity(new Identity("set_user_1", Optional.empty()))
+                .setRole("hive", new SelectedRole(SelectedRole.Type.ROLE, Optional.of("set_role_4")))
+                .build();
+
+        MaterializedResult actual = queryRunner.execute(unsetRole, "SELECT * FROM hive.information_schema.applicable_roles");
+        MaterializedResult expected = MaterializedResult.resultBuilder(unsetRole, createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType(), createUnboundedVarcharType())
+                .row("set_user_1", "USER", "admin", "NO")
```
Comment (cawallin) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: why is set_user_1 admin?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java [hunk 7]
```diff
@@ -323,6 +323,66 @@ public void testGrantRevokeRoleAccessControl()
...
+                .containsOnly(
+                        row("public"),
+                        row("role2"),
+                        row("role3"));
+
+        onPrestoAlice().executeQuery("SET ROLE role3");
+        QueryAssert.assertThat(onPrestoAlice().executeQuery("SELECT * FROM hive.information_schema.enabled_roles"))
+                .containsOnly(
+                        row("public"),
+                        row("role3"));
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java:501]: Like in the integration tests, you should drop the roles at the end

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java [hunk 8]
```diff
@@ -323,6 +323,66 @@ public void testGrantRevokeRoleAccessControl()
...
+    }
+
+    @Test(groups = {HIVE_CONNECTOR, ROLES, AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
+    public void testSetAdminRole()
+            throws Exception
+    {
+        QueryAssert.assertThat(onPresto().executeQuery("SELECT * FROM hive.information_schema.enabled_roles"))
+                .containsOnly(
+                        row("public"));
+        onPresto().executeQuery("SET ROLE admin");
```
Comment (cawallin) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java:511]: Is there a new connection created for every test? You probably should unset the role before the next test run, I think there was a provision in the standard for that if it's not possible with the commands you have now.
Also, how are these tests passing if SET ROLE isn't implemented for JDBC yet?

File: presto-client/src/main/java/com/facebook/presto/client/StatementClient.java [hunk 9]
```diff
@@ -148,6 +151,11 @@ private Request buildQueryRequest(ClientSession session, String query)
             builder.addHeader(PrestoHeaders.PRESTO_SESSION, entry.getKey() + "=" + entry.getValue());
         }
 
+        Map<String, SelectedRole> roles = session.getRoles();
```
Comment (cawallin) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClient.java:154]: Can you change the variable name or otherwise make it clear that this map is from catalog to role? (can be just in session.getRoles() javadoc or something)

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 10]
```diff
@@ -272,6 +298,23 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
...
+
+            ConnectorId connectorId = transactionManager.getOptionalCatalogMetadata(transactionId, catalogName)
+                    .orElseThrow(() -> new PrestoException(NOT_FOUND, "Catalog does not exist: " + catalogName))
+                    .getConnectorId();
+
+            if (role.getType() == SelectedRole.Type.ROLE) {
+                accessControl.checkCanSetRole(transactionId, identity, role.getRole().get(), catalogName);
+            }
+            roles.put(connectorId, role);
+            roles.put(createInformationSchemaConnectorId(connectorId), role);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/Session.java:317]: Why do you put these information schema connector roles?

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 11]
```diff
@@ -222,6 +234,17 @@ public Session createSession(
         return sessionProperties;
     }
 
+    private static Map<String, SelectedRole> parseRoleHeaders(HttpServletRequest servletRequest)
+    {
+        ImmutableMap.Builder<String, SelectedRole> roles = ImmutableMap.builder();
+        for (String header : splitSessionHeader(servletRequest.getHeaders(PRESTO_ROLE))) {
+            List<String> nameValue = Splitter.on('=').limit(2).trimResults().splitToList(header);
+            assertRequest(nameValue.size() == 2, "Invalid %s header", PRESTO_SESSION);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: s/PRESTO_SESSION/PRESTO_ROLE

**Review by cawallin — CHANGES_REQUESTED** [review]

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 1 above)
Comment (arhimondr) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:764]: It does. We do new line all across the presto code

File: presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java (hunk 2 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java:563]: During the "writing a doc" phase we have decided that access control is completely on connector. If we want to change this assumption - we need to change it everywhere. For how if you don't want some user to be an admin - you can revoke admin role for him in connector.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java (hunk 3 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:135]: This is kind of hack. InformationSchema connector receives `ConnectorSession`, but to be able to call Metadata methods it needs `Session`. So, Here we are re-creating session based on the connector session. And the identity in `Session` is not supposed to have any role assigned. Because we consider roles only on the connector level.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java (hunk 4 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:144]: I think this entire hack with session recreation needs to be removed at some point. In order to process roles we need to convert session from non transactional to transactional. It won't fail, becuase you cannot query without transaction.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java (hunk 5 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:412]: Yes. And it is completely connector specific. Connector is chosing what is the default ;-)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 6 above)
Comment (arhimondr) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: I changed that already. Now only `hive`, `hdfs` and `admin` users are admins when using file based hive metastore.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java (hunk 7 above)
Comment (arhimondr) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java:501]: It is done by `cleanup` method

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java (hunk 8 above)
Comment (arhimondr) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java:511]: > Is there a new connection created for every test?

Yes

> Also, how are these tests passing if SET ROLE isn't implemented for JDBC yet?

I've implemented it

File: presto-main/src/main/java/com/facebook/presto/Session.java (hunk 10 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/Session.java:317]: So we can get current role in information schema tables builder

File: presto-client/src/main/java/com/facebook/presto/client/StatementClient.java (hunk 9 above)
Comment (arhimondr) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClient.java:154]: > can be just in session.getRoles()

will do

File: presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java (hunk 2 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java:563]: OK, that's fine

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java (hunk 8 above)
Comment (cawallin) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java:511]: You should move the JDBC commit before this test commit then.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java (hunk 4 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:144]: What if we delay validating the roles, just like is done for connector properties? This is a pretty big hack and I don't think it would be acceptable upstream. (Because, e.g., I could see logic being added that doesn't allow you to start a transaction twice)

**Review by cawallin — APPROVED** [review]
I don't like the beginTransactionId part, and you should reorder the commits so that the test commit is after the commit where you add JDBC. Other than that, LGTM.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java (hunk 8 above)
Comment (arhimondr) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java:511]: It is before. It's just the way GitHub displays the order.
